sonify/core.py

```diff
@@ -123,6 +124,89 @@ def get_instrument(instrument_name):
             raise AttributeError('No instrument or percussion could be found by that name')
     return program_number - 1, instrument_type
 
+def auto_save_file(path):
+  directory, file_name=os.path.split(path)
+  while os.path.isfile(path):
+    pattern="(\d+)\)\."
+    if re.search(pattern, file_name) is None:
+      file_name=file_name.replace(".", "(0).")
+    else:
+      current_number=int(re.findall(pattern, file_name)[-1])
+      new_number=current_number + 1
+      file_name=file_name.replace(f"({current_number}).", f"({new_number}).")
+    if(directory==''):
+        path=file_name
+    else:
+        path=os.path.join(directory + os.sep + file_name)
+  return path   
+
+def save_to_midifile(input_data, track_type='single', key=None,number_of_octaves=4,output_path='',name='mymidifile'):
+    """
+    Save your midifile with this function instead of playing it!
+
+    data: (same list like play_to_midifile) list of tuples of x, y coordinates for pitch and timing
+          Optional: add a string to the start of the data list to specify instrument!
+    type: the type of data passed to create tracks. Either 'single' or 'multiple'
+    key: key to play back the graph -- see constants.py for current choices
+    number_of_octaves: number of octaves used to restrict the music playback
+                       when converting to a key
+    output_path: folder path where you want to save your midi files. Path like 'sounds/example/'. Always include the last '/'. 
+                 If you dont specify the path, files will be saved in the actual directory.
+    name = Include a name for your midi file, if name is not included, name will be "mymidifile"
+           If you use a loop and you don't include a name, output will be like: mymidifile.midi ,mymidifile(0).midi, mymidifile(1).midi ... (n files)   
+    """
+    
+    if key:
+        if track_type == 'multiple':
+            data = []
+            for data_list in input_data:
+                data.append(convert_to_key(data_list, key, number_of_octaves))
+        else:
+            data = convert_to_key(input_data, key, number_of_octaves)
+    else:
+        data = input_data
+    
+    
+    if track_type not in ['single', 'multiple']:
+        raise ValueError('Track type must be single or multiple')
+
+    if track_type == 'single':
+        data = [data]
```

Comment:
Hello @gregory-mc! Thank you SO much for the pull request! This functionality makes total sense! My only suggestion would be instead of copying and pasting a lot of the code from the `write_to_midifile` function, it might be nice to instead factor out the repeated bits so that any future change to how things work won't have to be made twice!

That being said, I'm also happy to merge as is and make the change later (if I can get to it!)

This was a fun hobby project, it's so cool to see other folks getting use out of it!!